<?xml version='1.0'?>

<launch>

<!--    
  <arg name="intersection" default="true" />
  <arg name="mcity" default="false" />
-->
  <arg name="single" default="true" />
  <arg name="use_rosbag" default="false" />
  <arg name="use_rviz" default="true" /> 
  <arg name="teleop_joy" default="true" /> 

  <!-- single car -->
  <include file="$(find prius_gazebo)/launch/single_car_intersection.launch" if="$(arg single)"/>

  <!-- multi car -->
  <include file="$(find prius_gazebo)/launch/multi_cars_intersection.launch" unless="$(arg single)"/>

  <!-- mcity world 
  <include file="$(find prius_gazebo)/launch/mcity.launch" if="$(arg mcity)"/>
-->

  <!-- joy stick teleop (1 or 2 cars) -->
  <include file="$(find prius_gazebo)/launch/includes/teleop_joy.xml"  if="$(arg teleop_joy)" >
    <arg name="single" value="$(arg single)" />
    <arg name="joy_maxVel" value="3" />
    <arg name="joy_accele" value="0.024"/>
  </include>
  
  <!-- keyboard teleop (only 1 car) -->
  <include file="$(find prius_gazebo)/launch/includes/teleop_key.xml"  unless="$(arg teleop_joy)" />
  
  <!-- gmapping 
  <include file="$(find solabot_navigation)/launch/includes/simulator/racecar_gmapping.xml"/>
-->


<!-- BEGIN OF LOCALIZATION -->

  <!-- Map server -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find prius_gazebo)/maps/intersection.yaml"/>

  <!-- localization of prius0 --> 
  <include file="$(find prius_gazebo)/launch/includes/localization.xml">
    <arg name="robot_ns" value="prius0" />
    <arg name="init_x" value="0.0" />
    <arg name="init_y" value="0.0" />
    <arg name="init_a" value="0.0" />
  </include>

  <!-- localization of prius1 --> 
  <include file="$(find prius_gazebo)/launch/includes/localization.xml" unless="$(arg single)" >
    <arg name="robot_ns" value="prius1" />
    <arg name="init_x" value="0.0" />
    <arg name="init_y" value="0.0" />
    <arg name="init_a" value="0.0" />
  </include>

  <!-- END OF LOCALIZATION -->

  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find prius_gazebo)/rviz/prius0_intersection_experiment.rviz" if="$(arg use_rviz)" />


  <!-- =============  rosbag  ============== -->
  <arg name="path" default="$(find prius_gazebo)/rosbag/20190509/" />
  <arg name="session" default="nothing.bag" />
  <arg name="topic1" default="/prius0/base_pose_ground_truth"/>
  <arg name="topic2" default="/prius1/base_pose_ground_truth"/>
  <arg name="topic3" default="/prius0/cmd_vel"/>
  <arg name="topic4" default="/prius1/cmd_vel"/>
  <arg name="topic5" default="/prius0/front_camera/image_raw/compressed"/>
  <arg name="topic6" default="/prius0/left_camera/image_raw/compressed"/>
  <arg name="topic7" default="/prius0/right_camera/image_raw/compressed"/>
  <arg name="topic8" default="/prius1/front_camera/image_raw/compressed"/>
  <arg name="topic9" default="/prius1/left_camera/image_raw/compressed"/>
  <arg name="topic10" default="/prius1/right_camera/image_raw/compressed"/>
  <node pkg="rosbag" type="record" name="rocorder" output="screen" args="-O $(arg path)$(arg session) $(arg topic1) $(arg topic2) $(arg topic3) $(arg topic4) $(arg topic5) $(arg topic6) $(arg topic7) $(arg topic8) $(arg topic9) $(arg topic10) " if="$(arg use_rosbag)"/>


</launch>
